### PR TITLE
mdata-client: add mdata-client release-20190228

### PR DESCRIPTION
**Pull request by jclulow**
This adds a binary package for the metadata client utilities that SmartOS HVM guests can use to get information from the hypervisor (or the cloud orchestration suite managing the hypervisor).

I built and installed this package in one of my OpenIndiana VMs running under KVM on a SmartOS host:

```
root@oi0:~# pkg set-publisher -g file:///ws/oi-userland/i386/repo userland
root@oi0:~# pkg install mdata-client
           Packages to install:  1
       Create boot environment: No
Create backup boot environment: No

DOWNLOAD                                PKGS         FILES    XFER (MB)   SPEED
Completed                                1/1           9/9      0.1/0.1    0B/s

PHASE                                          ITEMS
Installing new actions                         22/22
Updating package state database                 Done 
Updating package cache                           0/0 
Updating image state                            Done 
Creating fast lookup database                   Done 
Reading search index                            Done 
Updating search index                            1/1 

root@oi0:~# man mdata-list | head -20
MDATA-LIST(1m)                 Metadata Commands                MDATA-LIST(1m)






NAME
       mdata-list -- Fetch the list of available metadata key-value pairs.


SYNOPSIS
       /usr/sbin/mdata-list


DESCRIPTION
       The mdata-list command allows the user (or a script) to query the
       metadata for a guest instance running in a SmartDataCenter (SDC) cloud.
       Metadata values are generally set programmatically via CloudAPI, or
       during interactive provisioning via a Web Portal.  They may also be set

root@oi0:~# mdata-list

root@oi0:~# mdata-put date <<< $(date)

root@oi0:~# mdata-list
date

root@oi0:~# mdata-get date
March  2, 2019 at 02:05:10 PM PST

root@oi0:~# mdata-get sdc:alias
oi0

root@oi0:~# mdata-get sdc:server_uuid
36343330-3836-5553-4533-3238384d5053
```

I tried to make this look as much like the other Makefiles nearby as I could.  Please let me know if I need to fix anything!  Thanks.

### CODE REVIEW

**Comment by richlowe** [comment]
Probably should refer to it as Triton, not SDC.  I hear the cool kids do.

**Comment by jclulow** [comment]
> Probably should refer to it as Triton, not SDC. I hear the cool kids do.

Fixed.  Thanks.

**Comment by pyhalov** [comment]
Can you set/overwrite $(CC) so that userland compiler is certainly used, not just gcc? 
Also, please do "gmake sample-manifest" and add resulting file.

File: components/sysutils/mdata-client/Makefile [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+include $(WS_MAKE_RULES)/justmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+build:		$(BUILD_32)
+
+install:	$(INSTALL_32)
+
+test:		$(NO_TESTS)
+
+REQUIRED_PACKAGES += system/library
```
Comment (Mno-hime) [inline components/sysutils/mdata-client/Makefile:44]: Also run `gmake REQUIRED_PACKAGES`, so it's clear in the Makefile, what dependencied the build system is able to find on it's own. You'll get something like:
```Makefile
# Auto-generated dependencies
REQUIRED_PACKAGES += ...
```
Whatever is needed, but not found automatically, move above the comment.

Comment (jclulow) [reply components/sysutils/mdata-client/Makefile:44]: > Can you set/overwrite $(CC)

Done!  The software appears to build correctly, now using a fully qualified `/usr/gcc/6/bin/gcc`.

> please do "gmake sample-manifest" and add resulting file

I will do this, but it's not immediately clear to me what this file is for!  Can you point me to a document that explains it?

> run `gmake REQUIRED_PACKAGES`, so it's clear in the Makefile, what dependencied the build system is able to find on it's own

Done!  It's still just `system/library` but I now have the `# Auto-generated dependencies` banner.

Comment (Mno-hime) [reply components/sysutils/mdata-client/Makefile:44]: > > please do "gmake sample-manifest" and add resulting file
> 
> I will do this, but it's not immediately clear to me what this file is for! Can you point me to a document that explains it?

It's a good starting point for the actual mdata-client.p5m component file and on version update one can see via `git diff`, what new files were installed (or were removed) and decide if they should be put into the package.

E.g. see https://github.com/OpenIndiana/oi-userland/pull/4818/files. On version update three new files emerged in the sample manifest (reflects the install area) and then I decided to put them to the right component p5m file.

> Done! It's still just `system/library` but I now have the `# Auto-generated dependencies` banner.

That banner is a good sign that someone actually run the `REQUIRED_PACKAGES` target and that the list of required packages is more or less complete.

File: components/sysutils/mdata-client/Makefile [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2019 Joyent, Inc.
+#
+
+include ../../../make-rules/shared-macros.mk
```
Comment (Mno-hime) [inline components/sysutils/mdata-client/Makefile:16]: Adding `PREFERRED_BITS=64` before the include and changing common targets `s/32/64/` makes sure 64-bit binary is build and placed to the (usually 32-bit) area of `/usr/bin/`. E.g. see `components/archiver/gtar/` component.

Any reason to go with 32-bit binaries? We prefer 64-bit where possible, certainly for new stuff.

Comment (jclulow) [reply components/sysutils/mdata-client/Makefile:16]: At the moment, the software is only built and shipped 32-bit on any platform.  The `Makefile` doesn't currently offer a way to add different `CFLAGS` in a useful way.  I can try and fix that upstream, and then come back to improve the package here, but I'd like to get these binaries shipped as 32-bit first.

**Review by Mno-hime — APPROVED** [review]

**Pull request merged**